bootstrap: enable zlib for LLVM for Windows GNU

ognevny opened the pull request:
when compiling target LLVM with `opt-dist local` on Windows/MinGW, profraw files are being compressed with zlib, so compiling without it will make `llvm-profdata` complain about lacking of zlib support. this error is shown:

`profile uses zlib compression but the profile reader was built without zlib support`

example from llvm test suite: https://github.com/llvm/llvm-project/blob/76b5fcbf975547251faaeed8b567ea09d139a607/llvm/test/tools/llvm-profdata/nocompress.test#L15.

implementation of step 3. from https://rust-lang.zulipchat.com/#narrow/channel/122651-general/topic/experiment.20with.20.60opt-dist.20local.60.20in.20MSYS2/near/512393873

r? Kobzol
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

ognevny commented:
yeah, that doesn't work
`warning: D:/_/B/src/CLANG64/rustc-1.86.0-src/opt-artifacts\rustc-pgo/default_1016764578532249842_0_68.profraw: profile uses zlib compression but the profile reader was built without zlib support`

ognevny commented:
now it works, so ready :)

Kobzol commented on src/bootstrap/src/core/build_steps/llvm.rs:
```diff
@@ -370,8 +370,8 @@ impl Step for Llvm {
-        // Libraries for ELF section compression.
-        if !target.is_windows() {
+        // Libraries for ELF section compression and profraw files merging.
+        if !target.is_windows() || builder.config.rust_profile_generate.is_some() {
```
Hmm, this seems a bit random. We recompile LLVM several times during the opt-dist cycle, and this will enable zlib only when we specifically generate PGO profiles for Rust. What about when we generate PGO profiles for LLVM?

If Windows/GNU supports zlib just fine, then I think that we should make the condition more precise, and just allow zlib unconditionally for Windows/GNU, same as we do for other non-Windows OSes. Maybe just change the condition to `if !target.is_msvc()`?

ognevny replied:
> If Windows/GNU supports zlib just fine, then I think that we should make the condition more precise, and just allow zlib unconditionally for Windows/GNU, same as we do for other non-Windows OSes. Maybe just change the condition to `if !target.is_msvc()`?

yeah, that what I did locally at first

Kobzol replied:
I would expect the host and target to always be the same. Doing PGO with cross-compilation sounds quite difficult =D But I guess on some systems it might be viable.

ognevny committed:
bootstrap: enable zlib for LLVM when we compile it for profiling

when compiling target LLVM with `opt-dist local` on Windows/MinGW,
profraw files are being compressed with zlib, so compiling without it
will make `llvm-profdata` complain about lacking of zlib support. this
error is shown:

`profile uses zlib compression but the profile reader was built without
zlib support`

example from llvm test suite: https://github.com/llvm/llvm-project/blob/76b5fcbf975547251faaeed8b567ea09d139a607/llvm/test/tools/llvm-profdata/nocompress.test#L15

Kobzol commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit bf2696311be53bed4fe976bb25bed81f0d7477ce has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol bf2696311be53bed4fe976bb25bed81f0d7477ce -->
<!-- homu: {"type":"Approved","sha":"bf2696311be53bed4fe976bb25bed81f0d7477ce","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

mati865 commented:
This likely won't work because you need zlib library and headers on CI.
IIRC in the past it used to be enabled for windows-gnu, but when GHA images changed this broke the builds.

ognevny commented:
> This likely won't work because you need zlib library and headers on CI.

I'm sure that they are installed somewhere else. or they aren't?

mati865 commented:
At least not oficially: https://github.com/actions/runner-images/blob/main/images/windows/Windows2022-Readme.md

ognevny commented:
looked at CI build. CMake couldn't find Zlib, but build doesn't fail, so probably nothing changed for resulting LLVM
